improved filepackaging

DerThorsten opened the pull request:
This allows having a config file to specify what is `packaged` for which package.

patters can be of following types:
 * regex
 * file-extension
 * contains a substring
 
one can have a config as yaml
```yaml
packages:
  python-dateutil: 
    include_patterns:
      - type : regex
        pattern: '^(?!.*\/tests\/).*(.*.\.py$)|(.*.\.so$)|(.*dateutil-zoneinfo\.tar\.gz$)'
      - type: contains_any : 
        contains_any ['fubar','fu']
    exclude_patterns:
      - type: contains_any : 
        contains_any ['/debug/','/test/']

default: 
  include_patterns:
    - type : regex
      pattern: '^(?!.*\/tests\/).*(.*.\.py$)|(.*.\.so$)'
```

and we have a sensible default directly in the code

wolfv commented:
I really don't like complicated regex.

It seems that most of the examples could be expressed as straight-forward globs?

```
default: 
  include_patterns:
    - **/*.so
    - **/*.py
  exclude_patterns:
    - **/tests/*
```

DerThorsten commented:
I think we should allow support for regex since with them once can anything.
@wolfv  I will add support for `fnmatch` and make the defaults based on them

wolfv commented:
cool! 
Ok, let's keep regex. We can also identify a regex if the string starts with `^` and ends with `$`, then we could simplify the YAML.

DerThorsten commented:
@wolfv  I removed all the filters except the regex one and added the "glob"/"unix-wildcard-style" one.
I prefer to be explicit, so instead of relying on `^` to detect regex, we use the key `regex` for a regex and `pattern` for the UNIX style wildcard patterns.

The yaml now looks like:
``` yaml
packages:
  python-dateutil: 
    include_patterns:
      - regex: '^(?!.*\/tests\/).*(.*.\.py$)|(.*.\.so$)|(.*dateutil-zoneinfo\.tar\.gz$)'
      - patter: "*.so" 
      - patter: "**/include_me/*"
    exclude_patterns:
      - patter: "*.a" 
      - patter: "**/exclud_me/*"
default: 
  include_patterns:
    - regex: '^(?!.*\/tests\/).*(.*.\.py$)|(.*.\.so$)'
```

DerThorsten commented:
this PR is also needed to fix https://github.com/jupyterlite/xeus-python-kernel/issues/55#issuecomment-1195108548

martinRenou commented on empack/file_packager.py:
```diff
@@ -224,37 +159,55 @@ def emscripten_file_packager(
+    env_prefix: Path, outname, export_name, pack_outdir=None, download_emsdk=None
+):
+    # name of the env
+    print("env_prefix", env_prefix)
```
```suggestion
```

martinRenou commented on empack/file_packager.py:
```diff
@@ -224,37 +159,55 @@ def emscripten_file_packager(
 
-    lib_dir = env_prefix / "lib"
-    python_lib_dir = lib_dir / f"python{version}"
+    print(f"{env_name=} {env_root=}")
```
```suggestion
```

martinRenou commented on empack/file_packager.py:
```diff
@@ -224,37 +159,55 @@ def emscripten_file_packager(
+        "pack_python_core is deprecated, use `pack_environment`",
+        DeprecationWarning,
+    )
+    print(
+        "pack_python_core is deprecated, use `pack_environment`",
+    )
```
```suggestion
```

martinRenou reviewed (commented):
Thanks! There are some leftover prints

DerThorsten commented:
> Thanks! There are some leftover prints

thanks for catching

wolfv commented:
LGTM. I would squash the commits a bit more, but that's all.

DerThorsten commented:
we can squash everything, I dont mind
